### DISCUSSION TITLE
Refactoring of root project structure

### DISCUSSION THREAD

**Discussion by AndrewShkrob** [organicmaps/organicmaps#4710] [Ideas]
The current root folder of the project looks messy and depressing.
I’d suggest making refactoring and composing the common items into separate folders.

Here is an example:

```bash
├── .github
├── 3party
├── platform
│   ├── android
│   ├── iphone
│   └── desktop
├── cmake
├── data
│   ├── test_data
│   ├── styles
│   ├── **
│   ├── fonts
│   └── *.*
├── docs
├── tools
├── lib
│   ├── base
│   ├── coding
│   ├── cppjansson
│   ├── drape
│   ├── drape_frontend
│   ├── editor
│   ├── map
│   │   ├── src
│   │   │   └── map
│   │   │       ├── **/*.hpp
│   │   │       └── **/*.cpp
│   │   └── test
│   │       ├── **/*.hpp
│   │       └── **/*.cpp
│   ├── **
│   └── transit
├── *.*
└── .gitignore
```

#### Some comments regarding `lib/` folder

* I propose to separate tests from sources. From my pov, it looks much better this way. There are also some advantages of such an approach but all of them are subjective, so I won't bother you with my thoughts.
* You may notice the strange path to library sources: `lib/map/src/map/**/*.*`. It may look weird at first sight, but it's quite obvious and logical on the other hand. We can configure `cmake` in such a way to use libs in source code in the same way as it was before:

  ```cpp
  #include <map/framework.hpp>
  ```

  It won't break the existing codebase.
* If you don't like `lib/map/src/map/**/*.*`, we can try to make it in such a way: `lib/map/src/**/*.*`. I believe it's also possible via cmake.

#### Will these changes break the revision history?

No! And yes...

##### Why no?

Gihub supports `.git-blame-ignore-revs` that will help to skip commits with these changes in git-blame on github. So, nobody will notice these changes.
It's the same approach we used during android's package renaming [here](https://github.com/organicmaps/organicmaps/pull/3657).

##### Why yes?

`.git-blame-ignore-revs` doesn't work locally by default. People have to configure the local project ([docs](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file)).
I don't think it's a big problem. Many people don't use `git blame` locally and those who use can find a solution by themselves (I believe) or we can add this step to docs.

#### Advantages

* Modern and compact root folder structure.
* Help newcomers not to drown in the codebase.
* Easy to find the folder/module that you need without endless scrolling and searching.
* Easy to understand what "that" module stands for.
  Example: `openlr`.
  * What is it?
  * Looks like a 3party lib (it has `open` in the name, right?).
  * Why is it not in a `3party` folder? Is it a modified version of 3party lib?
  * *Checks the source code*
  * Hm... A lot of internal import. Looks like it's not a 3party... But why is it `open`?
  * *Googles `openlr`*
  * > Open standard for "procedures and formats for the encoding, transmission, and decoding of local data irrespective of the map" developed by TomTom. 
  * WTF? *Still don't understand what is it.*
  This example also illustrates the problem of lack of documentation :)
* These changes can be made iteratively, folder-by-folder/library-by-library
* Good place to review and remove files or tools that are not used now and can be removed.
* Good place to write basic documentation: "What this library is about?", "What problem does it solve" and so on.

#### Disadvantages and problems during implementation

* A lot of work
* Potentially can break different tools and scripts that depend on paths.
* Changes should be made carefully with a lot of testing: CI, tests, platfroms, tools + integration with your servers that build and distribute maps.

#### Disadvantages after implementation

* I don't see any

#### Your opinion

Please, tell me what you think about this idea.
If you agree to these changes, I can analyze the whole project and prepare the full new structure with all libs, tools reviewed, and so on. And we can discuss each of them in detail.

**Comment by Ferenc-** [comment]
I think `qt` would be a better fit under `platform` directly, I am not sure that `desktop` expresses the intent correctly, as `qt` is planned for convergent mobile friendly UI as well.
And there are `platform_linux.cpp` and `platform_mac.mm` which are not even written in the same language, so perhaps it would make sense to separate those?
